src/bipart.cpp

```diff
@@ -346,6 +372,86 @@ namespace libsemigroups {
       return result;
     }
 
+    void uniform_random(Bipartition& x) {
+      if (x.degree() == 0) {
+        return;
+      }
+      long double const N = 2 * x.degree();
+
+      static std::random_device rd;
+      static std::mt19937       mt(rd());
+
+      long double const c = M_E * bell_number(N);
+      if (std::isnan(c)) {
+        LIBSEMIGROUPS_EXCEPTION(
+            "the degree ({}) of the argument <x> (a bipartition) is too large, "
+            "please use random(x) instead",
+            N / 2);
+      }
+      std::discrete_distribution<uint64_t> g(
+          4 * N, 0, 4 * N, [&N, &c](long double m) -> long double {
+            long double num = std::pow(m, N);
+            long double den = c * factorial(m);
+            if (std::isnan(num) || std::isnan(den) || std::isnan(num / den)) {
+              LIBSEMIGROUPS_EXCEPTION("the degree ({}) of the argument <x> (a "
+                                      "bipartition) is too large, "
+                                      "please use random(x) instead",
+                                      N / 2);
+            }
+            return num / den;
+          });
```

Comment:
I've been looking at the sage implementation of this, and I think this constructor for `std::discrete_distribution` behaves slightly differently to sage's `GeneralDiscreteDistribution`.

From the C++ doc:
> Each of the weights is equal to $w_i$ = unary-op($x_\{min\}$ + δ(i + 0.5))  where $δ = \frac{(x_{max} − x_{min})}{count}$ and $i ∈ \\{0, ..., count − 1\\}$.

In particular, the weight of 0 will not be equal to 0 as it is in sage. I think it is this that is causing the issue with segfault outlined below